op-challenger/game/keccak/merkle/tree.go

```diff
@@ -0,0 +1,193 @@
+package merkle
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-challenger/game/keccak/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// BinaryMerkleTreeDepth is the depth of the merkle tree.
+const BinaryMerkleTreeDepth = 16
+
+var (
+	// MaxLeafCount is the maximum number of leaves in the merkle tree.
+	MaxLeafCount = 1<<BinaryMerkleTreeDepth - 1 // 2^16 - 1
+	// IndexOutOfBoundsError is returned when an index is out of bounds.
+	IndexOutOfBoundsError = errors.New("index out of bounds")
+	// zeroHashes is a list of empty hashes in the binary merkle tree, indexed by height.
+	zeroHashes [BinaryMerkleTreeDepth]common.Hash
+)
+
+func init() {
+	// Initialize the zero hashes. These hashes are pre-computed for the starting state of the tree, where all leaves
+	// are equal to `[32]byte{}`.
+	for height := 0; height < BinaryMerkleTreeDepth-1; height++ {
+		zeroHashes[height+1] = crypto.Keccak256Hash(zeroHashes[height][:], zeroHashes[height][:])
+	}
+}
+
+// MerkleProof is a keccak merkle proof, which is a list of hashes that prove the inclusion of a leaf in the merkle tree.
+type MerkleProof [][]byte
```

Comment:
nit: Given this is in the `merkle` package now (good choice) it can just be called `Proof` since it will then be `merkle.Proof`.